File: 24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        pre, p = None, head
        while p and p.next:
            if pre == None:
                head = p.next
                t = p.next.next
                p.next = t
                head.next = p
            else:
                pre.next = p.next
                t = p.next.next
                p.next = t
                pre.next.next = p
            pre = p
            p = p.next
        return head

File: go/src/leetcode/24. Swap Nodes in Pairs/24. Swap Nodes in Pairs.go
package main

type ListNode struct {
	Val int
	Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
	dummy := new(ListNode)
	p := dummy
	st := []*ListNode{}
	for head != nil {
		st = st[:0]
		tmp := head
		cnt := 2
		for ; tmp != nil && cnt != 0; tmp, cnt = tmp.Next, cnt-1 {
			st = append(st, tmp)
		}
		if cnt > 0 {
			p.Next = head
			return dummy.Next
		}

		for len(st) != 0 {
			p.Next = st[len(st)-1]
			st = st[:len(st)-1]
			p = p.Next
		}

		p.Next = tmp
		head = tmp
	}
	return dummy.Next
}
